[WPE] WPE Platform: Add input method context support for Wayland

zhani opened the pull request:
#### 77db03823160b5bc32d0088dcabd873449f77b7f
<pre>
[WPE] WPE Platform: Add input method context support for Wayland
<a href="https://bugs.webkit.org/show_bug.cgi?id=265649">https://bugs.webkit.org/show_bug.cgi?id=265649</a>

Reviewed by Carlos Garcia Campos.

Add WPE Platfrom default input method context support for wayland
text input v1 and v3.

* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkitWebViewConstructed):
* Source/WebKit/UIProcess/API/wpe/WebKitInputMethodContextImplWPE.cpp: Added.
(toWPEInputPurpose):
(toWPEInputHints):
(inputPurposeChangedCallback):
(inputHintsChangedCallback):
(wpeIMContextPreeditStartCallback):
(wpeIMContextPreeditChangedCallback):
(wpeIMContextPreeditEndCallback):
(wpeIMContextCommitCallback):
(wpeIMContextDeleteSurrounding):
(webkitInputMethodContextImplWPEGetPreedit):
(webkitInputMethodContextImplWPENotifyFocusIn):
(webkitInputMethodContextImplWPENotifyFocusOut):
(webkitInputMethodContextImplWPENotifyCursorArea):
(webkitInputMethodContextImplWPENotifySurrounding):
(webkitInputMethodContextImplWPEReset):
(webkit_input_method_context_impl_wpe_class_init):
(webkitInputMethodContextImplWPENew):
* Source/WebKit/UIProcess/API/wpe/WebKitInputMethodContextImplWPE.h: Added.
* Source/WebKit/WPEPlatform/CMakeLists.txt:
* Source/WebKit/WPEPlatform/wpe/WPEColor.cpp: Copied from Source/WebKit/WPEPlatform/wpe/wpe-platform.h.
(wpe_color_copy):
(wpe_color_free):
* Source/WebKit/WPEPlatform/wpe/WPEColor.h: Copied from Source/WebKit/WPEPlatform/wpe/WPEDisplayPrivate.h.
* Source/WebKit/WPEPlatform/wpe/WPEDisplay.cpp:
(wpeDisplayCreateInputMethodContext):
* Source/WebKit/WPEPlatform/wpe/WPEDisplay.h:
* Source/WebKit/WPEPlatform/wpe/WPEDisplayPrivate.h:
* Source/WebKit/WPEPlatform/wpe/WPEInputMethodContext.cpp: Added.
(_WPEInputMethodUnderline::_WPEInputMethodUnderline):
(wpe_input_method_underline_new):
(wpe_input_method_underline_copy):
(wpe_input_method_underline_free):
(wpe_input_method_underline_get_start_offset):
(wpe_input_method_underline_get_end_offset):
(wpe_input_method_underline_set_color):
(wpe_input_method_underline_get_color):
(wpeInputMethodContextGetProperty):
(wpe_input_method_context_class_init):
(wpe_input_method_context_new):
(wpe_input_method_context_get_view):
(wpe_input_method_context_get_display):
(wpe_input_method_context_get_preedit_string):
(wpe_input_method_context_focus_in):
(wpe_input_method_context_focus_out):
(wpe_input_method_context_set_cursor_area):
(wpe_input_method_context_set_surrounding):
(wpe_input_method_context_reset):
* Source/WebKit/WPEPlatform/wpe/WPEInputMethodContext.h: Added.
* Source/WebKit/WPEPlatform/wpe/wayland/CMakeLists.txt:
* Source/WebKit/WPEPlatform/wpe/wayland/WPEDisplayWayland.cpp:
(wpeDisplayWaylandDispose):
(wpeDisplayWaylandCreateInputMethodContext):
(wpeDisplayWaylandGetTextInputV1):
(wpeDisplayWaylandGetTextInputV3):
(wpe_display_wayland_class_init):
* Source/WebKit/WPEPlatform/wpe/wayland/WPEDisplayWaylandPrivate.h:
* Source/WebKit/WPEPlatform/wpe/wayland/WPEInputMethodContextWaylandV1.cpp: Added.
(textInputV1GetGlobalByContext):
(toTextInputV1Hints):
(toTextInputV1Purpose):
(truncateSurroundingTextIfeeded):
(textInputV1SetCursorRectangle):
(textInputV1SetSurroundingText):
(textInputV1SetContentType):
(textInputV1CommitState):
(textInputV1UpdateState):
(keysymModifiersGetMask):
(textInputV1GetGlobalByDisplay):
(wpeIMContextWaylandV1PurposeOrHintsChanged):
(wpeIMContextWaylandV1Constructed):
(wpeIMContextWaylandV1Dispose):
(wpeIMContextWaylandV1GetPreeditString):
(wpeIMContextWaylandV1FocusIn):
(wpeIMContextWaylandV1FocusOut):
(wpeIMContextWaylandV1SetCursorArea):
(wpeIMContextWaylandV1SetSurrounding):
(wpeIMContextWaylandV1Reset):
(wpe_im_context_wayland_v1_class_init):
(wpe_im_context_wayland_v1_new):
* Source/WebKit/WPEPlatform/wpe/wayland/WPEInputMethodContextWaylandV1.h: Copied from Source/WebKit/WPEPlatform/wpe/WPEDisplayPrivate.h.
* Source/WebKit/WPEPlatform/wpe/wayland/WPEInputMethodContextWaylandV3.cpp: Added.
(_Preedit::operator=):
(textInputV3GetGlobalByContext):
(toTextInputV3Hints):
(toTextInputV3Purpose):
(truncateSurroundingTextIfeeded):
(textInputV3SetCursorRectangle):
(textInputV3SetSurroundingText):
(textInputV3SetContentType):
(textInputV3CommitState):
(textInputV3UpdateState):
(textInputV3PreeditApply):
(textInputV3CommitApply):
(textInputV3DeleteSurroundingTextApply):
(textInputV3Enable):
(textInputV3Disable):
(textInputV3GetGlobalByDisplay):
(wpeIMContextWaylandV3PurposeOrHintsChanged):
(wpeIMContextWaylandV3Constructed):
(wpeIMContextWaylandV3Dispose):
(wpeIMContextWaylandV3GetPreeditString):
(wpeIMContextWaylandV3FocusIn):
(wpeIMContextWaylandV3FocusOut):
(wpeIMContextWaylandV3SetCursorArea):
(wpeIMContextWaylandV3SetSurrounding):
(wpeIMContextWaylandV3Reset):
(wpe_im_context_wayland_v3_class_init):
(wpe_im_context_wayland_v3_new):
* Source/WebKit/WPEPlatform/wpe/wayland/WPEInputMethodContextWaylandV3.h: Copied from Source/WebKit/WPEPlatform/wpe/WPEDisplayPrivate.h.
* Source/WebKit/WPEPlatform/wpe/wpe-platform.h:

Canonical link: <a href="https://commits.webkit.org/279725@main">https://commits.webkit.org/279725@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f815cae637b5f81c2f0632945affa4d611f83828

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/54206 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/33594 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/6751 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/57482 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/4930 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/41120 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/4881 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/43909 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/3305 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/56302 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/31857 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/46947 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25050 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/28667 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/3079 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/50367 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/4480 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/59075 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/29411 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/4648 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/51332 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/30583 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47055 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/50692 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/31553 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/8046 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/30369 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -881,6 +884,13 @@ static void webkitWebViewConstructed(GObject* object)
+        GRefPtr<WebKitInputMethodContext> imContext = adoptGRef(webkitInputMethodContextImplWPENew(priv->display.get()));
+        webkitInputMethodContextSetWebView(imContext.get(), webView);
+        priv->view->setInputMethodContext(imContext.get());
+        webkitInputMethodContextImplWPESetClientView(WEBKIT_INPUT_METHOD_CONTEXT_IMPL_WPE(imContext.get()), priv->view->wpeView());
```
I think set client view is needed in GTK because it needs to be set/unset when the widget is realized/unrealized. For WPE, the IM context will be attached to the same WPEView for its lifetime, so I think we can make `webkitInputMethodContextImplWPENew` receive a WPEView instead of a WPEDisplay

zhani replied:
fixed

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WebKitInputMethodContextImplWPE.cpp:
```diff
@@ -0,0 +1,253 @@
+    PROP_0,
+    PROP_DISPLAY,
+    N_PROPERTIES
+};
```
This is a private object, we don't need to use properties, it's simpler to just setup the context and everything in the New method after the `g_object_new`

zhani replied:
fixed

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WebKitInputMethodContextImplWPE.cpp:
```diff
@@ -0,0 +1,253 @@
+#include <wtf/glib/WTFGType.h>
+
+struct _WebKitInputMethodContextImplWPEPrivate {
+    GRefPtr<WPEDisplay> display;
```
I don't think we need to keep the display (or view if we change to use the view) since it's only used to create the WPE im context.

zhani replied:
fixed

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WebKitInputMethodContextImplWPE.cpp:
```diff
@@ -0,0 +1,253 @@
+        *cursorOffset = clampTo<unsigned>(offset);
+}
+
+static void webkitInputMethodContextImplWPENotifyFocusIn(WebKitInputMethodContext *context)
```
```suggestion
static void webkitInputMethodContextImplWPENotifyFocusIn(WebKitInputMethodContext* context)
```

zhani replied:
fixed

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WebKitInputMethodContextImplWPE.cpp:
```diff
@@ -0,0 +1,253 @@
+    wpe_im_context_focus_in(priv->context.get());
+}
+
+static void webkitInputMethodContextImplWPENotifyFocusOut(WebKitInputMethodContext *context)
```
```suggestion
static void webkitInputMethodContextImplWPENotifyFocusOut(WebKitInputMethodContext* context)
```

zhani replied:
fixed

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WebKitInputMethodContextImplWPE.cpp:
```diff
@@ -0,0 +1,253 @@
+    wpe_im_context_focus_out(priv->context.get());
+}
+
+static void webkitInputMethodContextImplWPENotifyCursorArea(WebKitInputMethodContext *context, int x, int y, int width, int height)
```
```suggestion
static void webkitInputMethodContextImplWPENotifyCursorArea(WebKitInputMethodContext* context, int x, int y, int width, int height)
```

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WebKitInputMethodContextImplWPE.h:
```diff
@@ -0,0 +1,37 @@
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Library General Public
+ * License as published by the Free Software Foundation; either
+Kokkola * version 2 of the License, or (at your option) any later version.
```
```suggestion
* version 2 of the License, or (at your option) any later version.
```

zhani replied:
Fixed.. don't know where that came from

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEColor.cpp:
```diff
@@ -0,0 +1,74 @@
+ *
+ * Returns: (transfer full): A copy of passed in #WPEColor.
+ */
+WPEColor* webkit_color_copy(WPEColor* color)
```
```suggestion
WPEColor* wpe_color_copy(WPEColor* color)
```

zhani replied:
fixed

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEColor.cpp:
```diff
@@ -0,0 +1,74 @@
+    fastFree(color);
+}
+
+G_DEFINE_BOXED_TYPE(WPEColor, wpe_color, webkit_color_copy, wpe_color_free);
```
```suggestion
G_DEFINE_BOXED_TYPE(WPEColor, wpe_color, wpe_color_copy, wpe_color_free);
```

zhani replied:
fixed

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEColor.cpp:
```diff
@@ -0,0 +1,74 @@
+ *
+ * Free the #WPEColor.
+ *
+ * Since: 2.24
```
Remove this

zhani replied:
fixed

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEColor.h:
```diff
@@ -0,0 +1,54 @@
+
+#define WPE_TYPE_COLOR (wpe_color_get_type())
+
+WPE_API GType wpe_color_get_type    (void);
```
```suggestion
WPE_API GType wpe_color_get_type (void);
```

zhani replied:
fixed

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEColor.h:
```diff
@@ -0,0 +1,54 @@
+#define WPE_TYPE_COLOR (wpe_color_get_type())
+
+WPE_API GType wpe_color_get_type    (void);
+WPE_API gboolean wpe_color_parse    (WPEColor *color, const gchar *color_string);
```
This is not in the implementation.

zhani replied:
fixed

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEColor.h:
```diff
@@ -0,0 +1,54 @@
+#ifndef WPEColor_h
+#define WPEColor_h
+
+#if !defined(__WPE_PLATFORM_H_INSIDE__) && !defined(BUILDING_WEBKIT)
+#error "Only <wpe/wpe-platform.h> can be included directly."
+#endif
```
This header should be included from wpe-platform.h

zhani replied:
Not sure what you mean?

carlosgcampos replied:
Add `#include <wpe/WPEColor.h>` in wpe-platform.h

zhani replied:
got it. Fixed

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEDisplay.cpp:
```diff
@@ -523,3 +523,9 @@ const char* wpe_display_get_drm_render_node(WPEDisplay* display)
     return wpeDisplayClass->get_drm_render_node ? wpeDisplayClass->get_drm_render_node(display) : nullptr;
 }
+
+WPEInputMethodContext* wpeDisplayCreateInputMethodContext(WPEDisplay* display)
```
It should be Get, not Create, we only want one IM context per display.

zhani replied:
We have multiple IM context one per view. Underneath we are just using one wl text input object but each wpe im context can have their own state which is restored when they get focus

zhani replied:
Adding clarification that this is now left as is for now

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEDisplay.h:
```diff
@@ -62,6 +63,8 @@ struct _WPEDisplayClass
     const char             *(* get_drm_device)                (WPEDisplay *display);
     const char             *(* get_drm_render_node)           (WPEDisplay *display);
 
+    WPEInputMethodContext   *(* create_input_method_context)    (WPEDisplay *display);
```
```suggestion
    WPEInputMethodContext  *(* get_input_method_context)   (WPEDisplay *display);
```

zhani replied:
We have multiple IM context one per view. Underneath we are just using one wl text input object but each wpe im context can have their own state which is restored when they get focus. Thus 'create' instead of 'get'

zhani replied:
Adding clarification that this is now left as is for now

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEInputMethodContext.cpp:
```diff
@@ -0,0 +1,516 @@
+ * Range of text in an preedit string to be shown underlined.
+ */
+
+G_DEFINE_BOXED_TYPE(WPEInputMethodUnderline, wpe_im_underline, wpe_im_underline_copy, wpe_im_underline_free)
```
We need to be consistent between class name and methods prefix, so it's either WPEIMUnderline and wpe_im_underline or WPEInputMethodUnderline and wpe_input_method_underline. I prefer the latter.

zhani replied:
Fixed

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEInputMethodContext.cpp:
```diff
@@ -0,0 +1,516 @@
+ */
+void wpe_im_underline_set_color(WPEInputMethodUnderline* underline, const WPEColor* color)
+{
+    if (color != nullptr) {
```
```suggestion
    if (!color) {
```

zhani replied:
Fixed

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEInputMethodContext.cpp:
```diff
@@ -0,0 +1,516 @@
+ * @underline: a #WPEInputMethodUnderline
+ * @color: (nullable): a #WPEColor or %NULL
+ *
+ * Set the color of the underline. If @rgba is %NULL the foreground text color will be used
```
```suggestion
 * Set the color of the underline. If @color is %NULL the foreground text color will be used
```

zhani replied:
fixed

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEInputMethodContext.cpp:
```diff
@@ -0,0 +1,516 @@
+    g_return_val_if_fail(WPE_IS_DISPLAY(display), nullptr);
+
+    return wpeDisplayCreateInputMethodContext(display);
+}
```
I don't think we should expose a constructor, we should always get the im context of the display

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEInputMethodContext.cpp:
```diff
@@ -0,0 +1,516 @@
+    /**
+     * WPEInputMethodContext:input-purpose:
+     *
+     * The purpose of the text field that the `WPEInputMethodContext is connected to.
```
```suggestion
     * The purpose of the text field that the #WPEInputMethodContext is connected to.
```

zhani replied:
fixed

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEInputMethodContext.cpp:
```diff
@@ -0,0 +1,516 @@
+}
+
+struct _WPEInputMethodContextPrivate {
+    GRefPtr<WPEDisplay> display;
```
Do we need to keep this here?

zhani replied:
get_display is used in many places so it's store as handy property and also it's used as a key to store global text input object on wayland impls. We could get display from WPEView also but it was more straighforward this way

carlosgcampos replied:
I'm worried about keeping an extra reference of the display here.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEInputMethodContext.cpp:
```diff
@@ -0,0 +1,516 @@
+ * @underlines: (out) (transfer full) (nullable) (element-type WPEInputMethodUnderline): location to store the underlines as a #GList of #WPEInputMethodUnderline
+ * @cursor_offset: (out) (nullable): location to store the position of cursor in preedit string
+ *
+ *  Get the pre-edit string and a list of WPEInputMethodUnderline.
```
```suggestion
 * Get the pre-edit string and a list of WPEInputMethodUnderline.
```

zhani replied:
Fixed

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEInputMethodContext.h:
```diff
@@ -0,0 +1,176 @@
+#ifndef WPEInputMethodContext_h
+#define WPEInputMethodContext_h
+
+#if !defined(__WPE_PLATFORM_H_INSIDE__) && !defined(BUILDING_WEBKIT)
+#error "Only <wpe/wpe-platform.h> can be included directly."
+#endif
```
This header should be included from wpe-platform.h

zhani replied:
Not sure what you mean?

carlosgcampos replied:
Add `#include <wpe/WPEInputMethodContext.h>` to wpe-platform.h

zhani replied:
Fixed

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEInputMethodContext.h:
```diff
@@ -0,0 +1,176 @@
+};
+
+WPE_API WPEInputMethodContext   *wpe_im_context_new                     (WPEDisplay            *display);
+WPE_API void                     wpe_im_context_set_client_view         (WPEInputMethodContext *context, WPEView *view);
```
Use multiple lines for the parameters and leave only on space between function name and ( for the longer function and them line up the rest with that.

zhani replied:
fixed

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEDisplayWayland.cpp:
```diff
@@ -209,6 +224,13 @@ const struct wl_registry_listener registryListener = {
         else if (!std::strcmp(interface, "zwp_linux_dmabuf_v1"))
             priv->linuxDMABuf = static_cast<struct zwp_linux_dmabuf_v1*>(wl_registry_bind(registry, name, &zwp_linux_dmabuf_v1_interface, std::min<uint32_t>(version, 4)));
+        else if (!std::strcmp(interface, "zwp_text_input_manager_v1")) {
+            priv->textInputManagerV1 = static_cast<struct zwp_text_input_manager_v1*>(wl_registry_bind(registry, name, &zwp_text_input_manager_v1_interface, 1));
+            priv->textInputV1 = zwp_text_input_manager_v1_create_text_input(priv->textInputManagerV1);
+        } else if (!std::strcmp(interface, "zwp_text_input_manager_v3")) {
+            priv->textInputManagerV3 = static_cast<struct zwp_text_input_manager_v3*>(wl_registry_bind(registry, name, &zwp_text_input_manager_v3_interface, 1));
+            priv->textInputV3 = zwp_text_input_manager_v3_get_text_input(priv->textInputManagerV3, priv->wlSeat->seat());
```
We could create the WPEInputMethodContext here passing it the zwp_text_input_, similar to what we do with WPEMonitor, for example

zhani replied:
we can have multiple im context per view with their own state that needs to be restored on focus

zhani replied:
if we create WPEInputMethodContext here it means that we would have only on global WPEInputMethodContext and that's not the intention but we want to have muliple WPEInputMethodContext with their own state which is restored to global text input object on each focus

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEDisplayWayland.cpp:
```diff
@@ -375,6 +397,20 @@ static WPEView* wpeDisplayWaylandCreateView(WPEDisplay* display)
+    if (!priv->wlDisplay || !priv->wlCompositor)
+        return nullptr;
+
+    if (priv->textInputManagerV3)
+        return wpe_im_context_wayland_v3_new(WPE_DISPLAY_WAYLAND(display));
+    if (priv->textInputManagerV1)
+        return wpe_im_context_wayland_v1_new(WPE_DISPLAY_WAYLAND(display));
```
This should just return the created context, not create a new one on every call, I think it's simpler if we have a single context attached to the display. In GTK things works differently because the IM context is a module loaded at runtime.

zhani replied:
we can have multiple im context per view with their own state that needs to be restored on focus

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEInputMethodContextWaylandV1.cpp:
```diff
@@ -0,0 +1,680 @@
+typedef struct _TextInputV1Global TextInputV1Global;
+
+struct _TextInputV1Global {
+    struct zwp_text_input_v1* textInput;
```
This should be in the private struct, since the context is already global

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEInputMethodContextWaylandV1.cpp:
```diff
@@ -0,0 +1,680 @@
+        int32_t y;
+        int32_t width;
+        int32_t height;
+    } cursor_rect;
```
```suggestion
    } cursorRect;
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEInputMethodContextWaylandV1.cpp:
```diff
@@ -0,0 +1,680 @@
+    struct {
+        int32_t index;
+        uint32_t length;
+    } pending_surrounding_delete;
```
```suggestion
    } pendingSurroundingDelete;
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEInputMethodContextWaylandV1.cpp:
```diff
@@ -0,0 +1,680 @@
+    zwp_text_input_v1_set_surrounding_text(global->textInput,
+        truncatedText ? truncatedText : context->priv->surrounding.text,
+        cursorIndex, anchorIndex);
+    g_free(truncatedText);
```
Use GUniquePtr

zhani replied:
Fixed

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEInputMethodContextWaylandV1.cpp:
```diff
@@ -0,0 +1,680 @@
+
+    g_clear_object(&self->priv->view);
+    g_free(self->priv->preedit.text);
+    g_free(self->priv->surrounding.text);
```
Use GUniquePtr or CString

zhani replied:
There's also GList pointer for underlines which I just noticed that is not handled by finalize. I think that's something that should be freed explicitly even after changing text char* pointers to GUniquePtr?

zhani replied:
Fixed

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEInputMethodContextWaylandV1.cpp:
```diff
@@ -0,0 +1,680 @@
+{
+    GObjectClass* objectClass = G_OBJECT_CLASS(klass);
+    objectClass->constructed = wpeIMContextWaylandV1Constructed;
+    objectClass->finalize = wpeIMContextWaylandV1Finalize;
```
Use dispose, finalize is already set by WEBKIT_DEFINE_TYPE to call the private struct destructor

zhani replied:
fixed

zhani replied:
I was not aware of this but why it's not ok to use finalize if I call parent finalize anyways?

carlosgcampos replied:
WEBKIT_DEFINE_TYPE macro already adds a finalize implementation that calls the private struct destructor. That allows us to use C++ types and smart pointers in GObject private structs allocated by glib. If you add your own the one defined by the macro is not used and the destructor not called.

zhani replied:
Fixed

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEInputMethodContextWaylandV3.cpp:
```diff
@@ -0,0 +1,628 @@
+        int32_t y;
+        int32_t width;
+        int32_t height;
+    } cursor_rect;
```
```suggestion
    } cursorRect;
```

zhani replied:
fixed

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEInputMethodContextWaylandV3.cpp:
```diff
@@ -0,0 +1,628 @@
+    struct {
+        uint32_t beforeLength;
+        uint32_t afterLength;
+    } pending_surrounding_delete;
```
```suggestion
    } pendingSurroundingDelete;
```

zhani replied:
fixed

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEInputMethodContextWaylandV3.cpp:
```diff
@@ -0,0 +1,628 @@
+    } pending_surrounding_delete;
+};
+
+WEBKIT_DEFINE_FINAL_TYPE(WPEIMContextWaylandV3, wpe_im_context_wayland_v3, WPE_TYPE_INPUT_METHOD_CONTEXT, WPEInputMethodContext)
```
I think there's some common code betwen v1 and v3, so maybe we can design this differently to avoid code duplication.

zhani replied:
There are similarities. My thought was that it's better to keep them separated as they are unstable and a bit vague protocols

carlosgcampos reviewed (changes requested):
I couldn't review everything in detail, but I have added enough comments for a new iteration.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEInputMethodContextWaylandV1.h:
```diff
@@ -0,0 +1,46 @@
+#define WPE_TYPE_IM_CONTEXT_WAYLAND_V1 (wpe_im_context_wayland_v1_get_type())
+WPE_API G_DECLARE_FINAL_TYPE (WPEIMContextWaylandV1, wpe_im_context_wayland_v1, WPE, IM_CONTEXT_WAYLAND_V1, WPEInputMethodContext)
+
+WPE_API WPEInputMethodContext   *wpe_im_context_wayland_v1_new (WPEDisplayWayland *display);
```
I forgot to say, this shouldn't be public, I think. Same for the V3.

zhani replied:
fixed

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEColor.cpp:
```diff
@@ -0,0 +1,72 @@
+    fastFree(color);
+}
+
+G_DEFINE_BOXED_TYPE(WPEColor, wpe_color, wpe_color_copy, wpe_color_free);
```
Move this right after the initial section documentation and remove the trailing semicolon.

zhani replied:
This follows similar structure of code as in WPERectangle.cpp. If I move this then I need declare wpe_color_copy and wpe_color_free above G_DEFINE_BOXED_TYPE line. Removed semicolon but waiting for response to my note

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WebKitInputMethodContextImplWPE.cpp:
```diff
@@ -0,0 +1,198 @@
+
+#include "config.h"
+#include "WebKitInputMethodContextImplWPE.h"
+
```
`#if ENABLE(WPE_PLATFORM)`

zhani replied:
Fixed

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEInputMethodContext.cpp:
```diff
@@ -0,0 +1,516 @@
+
+struct _WPEInputMethodUnderline {
+    _WPEInputMethodUnderline(unsigned startOffset, unsigned endOffset)
+        : startOffset(startOffset), endOffset(endOffset)
```
```suggestion
        : startOffset(startOffset)
        , endOffset(endOffset)
```

zhani replied:
fixed

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEInputMethodContext.cpp:
```diff
@@ -0,0 +1,516 @@
+ * Set the client view for the input context; this view is usually tracked
+ * for focus lost and may also be used for purposes internal to the input
+ * method.
+ **/
```
```suggestion
 */
```

zhani replied:
fixed

carlosgcampos commented on Source/WebKit/UIProcess/API/wpe/WebKitInputMethodContextImplWPE.cpp:
```diff
@@ -0,0 +1,202 @@
+
+    auto* priv = WEBKIT_INPUT_METHOD_CONTEXT_IMPL_WPE(context)->priv;
+    priv->context = adoptGRef(wpe_input_method_context_new(wpe_view_get_display(view)));
+    wpe_input_method_context_set_client_view(priv->context.get(), view);
```
I'm not sure we need this on wpe since we always set it on construction and we never set it to NULL.

zhani replied:
If I understand this now correctly, you are suggesting that view instance should be passed in wpe_input_method_context_new instead of passing display? Meaning changing line 190 as follows
```diff
-priv->context = adoptGRef(wpe_input_method_context_new(wpe_view_get_display(view)));
+priv->context = adoptGRef(wpe_input_method_context_new(view));
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/WPEInputMethodContext.h:
```diff
@@ -0,0 +1,190 @@
+WPE_API guint                    wpe_input_method_underline_get_end_offset   (WPEInputMethodUnderline *underline);
+WPE_API void                     wpe_input_method_underline_set_color        (WPEInputMethodUnderline *underline,
+                                                                              const WPEColor          *color);
+WPE_API const WPEColor*          wpe_input_method_underline_get_color        (WPEInputMethodUnderline *underline);
```
```suggestion
WPE_API const WPEColor          *wpe_input_method_underline_get_color        (WPEInputMethodUnderline *underline);
```

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEInputMethodContextWaylandV1.h:
```diff
@@ -0,0 +1,46 @@
+
+#if !defined(__WPE_WAYLAND_H_INSIDE__) && !defined(BUILDING_WEBKIT)
+#error "Only <wpe/wayland/wpe-wayland.h> can be included directly."
+#endif
```
This shouldn't be public

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEInputMethodContextWaylandV1.h:
```diff
@@ -0,0 +1,46 @@
+G_BEGIN_DECLS
+
+#define WPE_TYPE_IM_CONTEXT_WAYLAND_V1 (wpe_im_context_wayland_v1_get_type())
+WPE_API G_DECLARE_FINAL_TYPE (WPEIMContextWaylandV1, wpe_im_context_wayland_v1, WPE, IM_CONTEXT_WAYLAND_V1, WPEInputMethodContext)
```
This shouldn't be public

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEInputMethodContextWaylandV3.h:
```diff
@@ -0,0 +1,46 @@
+
+#if !defined(__WPE_WAYLAND_H_INSIDE__) && !defined(BUILDING_WEBKIT)
+#error "Only <wpe/wayland/wpe-wayland.h> can be included directly."
+#endif
```
This shouldn't be public

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEInputMethodContextWaylandV3.h:
```diff
@@ -0,0 +1,46 @@
+G_BEGIN_DECLS
+
+#define WPE_TYPE_IM_CONTEXT_WAYLAND_V3 (wpe_im_context_wayland_v3_get_type())
+WPE_API G_DECLARE_FINAL_TYPE (WPEIMContextWaylandV3, wpe_im_context_wayland_v3, WPE, IM_CONTEXT_WAYLAND_V3, WPEInputMethodContext)
```
This shouldn't be public

zhani committed:
[WPE] WPE Platform: Add input method context support for Wayland
https://bugs.webkit.org/show_bug.cgi?id=265649

Reviewed by Carlos Garcia Campos.

Add WPE Platfrom default input method context support for wayland
text input v1 and v3.

* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkitWebViewConstructed):
* Source/WebKit/UIProcess/API/wpe/WebKitInputMethodContextImplWPE.cpp: Added.
(toWPEInputPurpose):
(toWPEInputHints):
(inputPurposeChangedCallback):
(inputHintsChangedCallback):
(wpeIMContextPreeditStartCallback):
(wpeIMContextPreeditChangedCallback):
(wpeIMContextPreeditEndCallback):
(wpeIMContextCommitCallback):
(wpeIMContextDeleteSurrounding):
(webkitInputMethodContextImplWPEGetPreedit):
(webkitInputMethodContextImplWPENotifyFocusIn):
(webkitInputMethodContextImplWPENotifyFocusOut):
(webkitInputMethodContextImplWPENotifyCursorArea):
(webkitInputMethodContextImplWPENotifySurrounding):
(webkitInputMethodContextImplWPEReset):
(webkit_input_method_context_impl_wpe_class_init):
(webkitInputMethodContextImplWPENew):
* Source/WebKit/UIProcess/API/wpe/WebKitInputMethodContextImplWPE.h: Added.
* Source/WebKit/WPEPlatform/CMakeLists.txt:
* Source/WebKit/WPEPlatform/wpe/WPEColor.cpp: Copied from Source/WebKit/WPEPlatform/wpe/wpe-platform.h.
(wpe_color_copy):
(wpe_color_free):
* Source/WebKit/WPEPlatform/wpe/WPEColor.h: Copied from Source/WebKit/WPEPlatform/wpe/WPEDisplayPrivate.h.
* Source/WebKit/WPEPlatform/wpe/WPEDisplay.cpp:
(wpeDisplayCreateInputMethodContext):
* Source/WebKit/WPEPlatform/wpe/WPEDisplay.h:
* Source/WebKit/WPEPlatform/wpe/WPEDisplayPrivate.h:
* Source/WebKit/WPEPlatform/wpe/WPEInputMethodContext.cpp: Added.
(_WPEInputMethodUnderline::_WPEInputMethodUnderline):
(wpe_input_method_underline_new):
(wpe_input_method_underline_copy):
(wpe_input_method_underline_free):
(wpe_input_method_underline_get_start_offset):
(wpe_input_method_underline_get_end_offset):
(wpe_input_method_underline_set_color):
(wpe_input_method_underline_get_color):
(wpeInputMethodContextGetProperty):
(wpe_input_method_context_class_init):
(wpe_input_method_context_new):
(wpe_input_method_context_get_view):
(wpe_input_method_context_get_display):
(wpe_input_method_context_get_preedit_string):
(wpe_input_method_context_focus_in):
(wpe_input_method_context_focus_out):
(wpe_input_method_context_set_cursor_area):
(wpe_input_method_context_set_surrounding):
(wpe_input_method_context_reset):
* Source/WebKit/WPEPlatform/wpe/WPEInputMethodContext.h: Added.
* Source/WebKit/WPEPlatform/wpe/wayland/CMakeLists.txt:
* Source/WebKit/WPEPlatform/wpe/wayland/WPEDisplayWayland.cpp:
(wpeDisplayWaylandDispose):
(wpeDisplayWaylandCreateInputMethodContext):
(wpeDisplayWaylandGetTextInputV1):
(wpeDisplayWaylandGetTextInputV3):
(wpe_display_wayland_class_init):
* Source/WebKit/WPEPlatform/wpe/wayland/WPEDisplayWaylandPrivate.h:
* Source/WebKit/WPEPlatform/wpe/wayland/WPEInputMethodContextWaylandV1.cpp: Added.
(textInputV1GetGlobalByContext):
(toTextInputV1Hints):
(toTextInputV1Purpose):
(truncateSurroundingTextIfeeded):
(textInputV1SetCursorRectangle):
(textInputV1SetSurroundingText):
(textInputV1SetContentType):
(textInputV1CommitState):
(textInputV1UpdateState):
(keysymModifiersGetMask):
(textInputV1GetGlobalByDisplay):
(wpeIMContextWaylandV1PurposeOrHintsChanged):
(wpeIMContextWaylandV1Constructed):
(wpeIMContextWaylandV1Dispose):
(wpeIMContextWaylandV1GetPreeditString):
(wpeIMContextWaylandV1FocusIn):
(wpeIMContextWaylandV1FocusOut):
(wpeIMContextWaylandV1SetCursorArea):
(wpeIMContextWaylandV1SetSurrounding):
(wpeIMContextWaylandV1Reset):
(wpe_im_context_wayland_v1_class_init):
(wpe_im_context_wayland_v1_new):
* Source/WebKit/WPEPlatform/wpe/wayland/WPEInputMethodContextWaylandV1.h: Copied from Source/WebKit/WPEPlatform/wpe/WPEDisplayPrivate.h.
* Source/WebKit/WPEPlatform/wpe/wayland/WPEInputMethodContextWaylandV3.cpp: Added.
(_Preedit::operator=):
(textInputV3GetGlobalByContext):
(toTextInputV3Hints):
(toTextInputV3Purpose):
(truncateSurroundingTextIfeeded):
(textInputV3SetCursorRectangle):
(textInputV3SetSurroundingText):
(textInputV3SetContentType):
(textInputV3CommitState):
(textInputV3UpdateState):
(textInputV3PreeditApply):
(textInputV3CommitApply):
(textInputV3DeleteSurroundingTextApply):
(textInputV3Enable):
(textInputV3Disable):
(textInputV3GetGlobalByDisplay):
(wpeIMContextWaylandV3PurposeOrHintsChanged):
(wpeIMContextWaylandV3Constructed):
(wpeIMContextWaylandV3Dispose):
(wpeIMContextWaylandV3GetPreeditString):
(wpeIMContextWaylandV3FocusIn):
(wpeIMContextWaylandV3FocusOut):
(wpeIMContextWaylandV3SetCursorArea):
(wpeIMContextWaylandV3SetSurrounding):
(wpeIMContextWaylandV3Reset):
(wpe_im_context_wayland_v3_class_init):
(wpe_im_context_wayland_v3_new):
* Source/WebKit/WPEPlatform/wpe/wayland/WPEInputMethodContextWaylandV3.h: Copied from Source/WebKit/WPEPlatform/wpe/WPEDisplayPrivate.h.
* Source/WebKit/WPEPlatform/wpe/wpe-platform.h:

Canonical link: https://commits.webkit.org/279725@main
